### ISSUE TITLE
Error when deserializing untagged enum

### ISSUE THREAD

**Issue by alex-shapiro** [3Hren/msgpack-rust#148]
Deserialization fails for any untagged enum whose variant contains an enum in its associated data. Example:

```rust
#[derive(Serialize, Deserialize)]
#[serde(untagged)]
enum Foo {
    A(Bar),
}

#[derive(Serialize, Deserialize)]
enum Bar {
    A,
}

let enum1 = Foo::A(Bar::A);
let bytes = rmp_serde::to_vec(&enum1).unwrap();
let enum2 = rmp_serde::from_slice::<Foo>(&bytes).unwrap();
```

Serializing produces the expected value, `[146, 0, 144]` (JSON equivalent is `[0, []]`). 
Deserializing fails with the error `Syntax("data did not match any variant of untagged enum Foo")'`.
The problem appears specific to `rmp_serde` (the above example works with `serde_json`).

My current workaround is to manually implement Deserialize, but this isn't ideal:

```rust
impl<'de> serde::Deserialize<'de> for Foo {
    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error> where D: serde::Deserializer<'de> {
        if let Ok(bar) = Bar::deserialize(deserializer) {
            Ok(Foo::A(bar))
        } else {
            Err(D::Error::custom("data did not match any variant of untagged enum Foo"))
        }
    }
}
```

**Comment by alex-shapiro** [comment]
After further digging, the problem is that `rmp_serde`'s enum representation does not conform to serde's expectations.

When deserializing an untagged enum, serde first deserializes to an intermediate representation called Content. The Content representation is then deserialized by ContentDeserializer. 

ContentDeserializer expects an enum to be encoded in one of two ways:
* as a String (for a unit variant)
* as a Map, where the key is the variant name and the value is the associated data

rmp_serde encodes enums as a Seq, where the first element is the variant name and the second element is a Seq of the associated data. This means serde fails when it tries to decode the Content-encoded enum: https://github.com/serde-rs/serde/blob/master/serde/src/private/de.rs#L1104.

I see a two solutions:

* update `ContentDeserializer` in serde to accept Seq-represented enums.
* update `rmp_serde`'s enum representation from Seq to Map (e.g. `[146, 0, 144]` becomes `[129, 0, 144]`). This has the benefit of bringing rmp_serde in line with Serde's expectations, but it is *not* backwards-compatible.

**Comment by 3Hren** [comment]
(De)serializing as a map seems the less of those evils, because:
- It does not change the result size.
- MessagePack does not specify how to deal with such cases, so it's language-dependent trouble.
- We've already broken backward compatibility by changing the way newtypes are handled.

**Comment by alex-shapiro** [comment]
@3Hren I agree. I started a [fork to update enum representation](https://github.com/3Hren/msgpack-rust/compare/master...alex-shapiro:master). Right now untagged enums can be decoded if they have nested unit, newtype, or tuple enum variants. 

Struct enum variants don't work yet because ContentDeserializer expects struct variant data to be represented as a map: https://github.com/serde-rs/serde/blob/4773863e3a70196a7ad0a75aafebc89a4870fcf8/serde/src/private/de.rs#L1631. I'm looking into some workarounds.

**Comment by alex-shapiro** [comment]
fixed in #149
